### PR TITLE
Allow mean return nan for 0dim

### LINKED ISSUE #888: torch.mean with empty dim should nan
```
import torch
import torch_xla

device = 'xla:0'
# device='cpu'

shape = (2, 0, 4)
x = torch.randn(shape, device=device)
y = torch.mean(x)
print(y)
```
Expected behavior: cpu produce nan while xla produce 0.

### PR DESCRIPTION
**Pull request by ailzhang**
Also updated tests which covers most reduction ops. 
fixes #888

### CODE REVIEW

File: torch_xla/csrc/reduction.cpp [hunk 1]
```diff
@@ -31,7 +31,9 @@ ReductionInfo GetReductionInfo(
       rinfo.new_dimensions.push_back(shape.dimensions(i));
     }
   }
-  XLA_CHECK_GT(rinfo.element_count, 0);
+  if (check_positive_numel) {
```
Comment (dlibenzi) [inline torch_xla/csrc/reduction.cpp]: This should not be here IMHO.
I mean, we end up adding one line anyway at call site, so be better do the check over there.

Comment (ailzhang) [reply torch_xla/csrc/reduction.cpp]: Getting element_count in each call site wasn't free :P Or am I missing an API in `xla::Shape` that can do it?

Comment (dlibenzi) [reply torch_xla/csrc/reduction.cpp]: It's part of `rinfo`, isn't it?

File: torch_xla/csrc/reduction.cpp [hunk 2]
```diff
@@ -73,7 +72,7 @@ xla::XlaOp CreateSummation(
   xla::XlaOp result = xla::Reduce(
       input, init_value, XlaHelpers::CreateAddComputation(shape.element_type()),
       dimensions);
-  if (scale && rinfo.element_count > 1) {
```
Comment (dlibenzi) [inline torch_xla/csrc/reduction.cpp]: Are you sure?
Coming in with count==0 means division by zero below...

Comment (ailzhang) [reply torch_xla/csrc/reduction.cpp]: yea `nan` is the expected input for this case: in the sense that numpy and pytorch both returns nan...
```
In [1]: import numpy as np

In [2]: import torch

In [3]: a = torch.rand(2, 0, 4)

In [4]: a.mean()
Out[4]: tensor(nan)

In [5]: b = np.random.rand(2, 0, 4)

In [6]: b.mean()
/home/ubuntu/miniconda3/envs/maskrcnn36/bin/ipython:1: RuntimeWarning: Mean of empty slice.
  #!/home/ubuntu/miniconda3/envs/maskrcnn36/bin/python
/home/ubuntu/miniconda3/envs/maskrcnn36/lib/python3.6/site-packages/numpy/core/_methods.py:85: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
Out[6]: nan
```

Comment (dlibenzi) [reply torch_xla/csrc/reduction.cpp]: What about the line:
```Python
1.0f / static_cast<float>(rinfo.element_count)
```

?

Comment (ailzhang) [reply torch_xla/csrc/reduction.cpp]: sorry I didn't get this - what's the issue about that line? It produces a nan which is then multiplied to result.

Comment (dlibenzi) [reply torch_xla/csrc/reduction.cpp]: Division by zero is Undefined Behavior, and many times depends on compiler options (whether to SIGFPE or not).
So better not rely on it returning NaN.

Comment (ailzhang) [reply torch_xla/csrc/reduction.cpp]: ehhh what's a better way to get nan in this case?

Comment (asuhan) [reply torch_xla/csrc/reduction.cpp]: Not sure what the most reliable way of construction NaN in XLA. I assume `ConstantR0` with a NaN float value should work.

Comment (dlibenzi) [reply torch_xla/csrc/reduction.cpp]: ScalarValue already does that.
I am not sure how XLA "feels" about NaN uploaded to TPU device. We can check.
In any case, that should be:
```C++
rinfo.element_count > 0 ? 1.0f / static_cast<float>(rinfo.element_count) : NAN
```

File: torch_xla/csrc/reduction.cpp [hunk 3]
```diff
@@ -74,8 +74,9 @@ xla::XlaOp CreateSummation(
       dimensions);
   if (scale) {
     xla::XlaOp scale = XlaHelpers::ScalarValue<float>(
-        1.0f / static_cast<float>(rinfo.element_count), shape.element_type(),
-        input.builder());
+        rinfo.element_count > 0 ? 1.0f / static_cast<float>(rinfo.element_count)
+                                : NAN,
```
Comment (dlibenzi) [inline torch_xla/csrc/reduction.cpp]: I'd `#include <cmath>` and run `clang-format`.

Comment (ailzhang) [reply torch_xla/csrc/reduction.cpp]: Hmmm `clang-format` seems to be happy with this line ;) 
Btw somehow it builds and passes test without cmath...hmmm but I just added anyway

Comment (dlibenzi) [reply torch_xla/csrc/reduction.cpp]: Oh, I know it compiles. But relies on some other include to pull the proper include, which is wrong.
If that include will stop including cmath, we will be suddenly breaking.
